homeassistant/components/esphome/__init__.py

```diff
@@ -140,34 +140,43 @@ def async_on_service_call(service: HomeassistantServiceCall) -> None:
                 )
             )
 
-    async def send_home_assistant_state_event(event: Event) -> None:
-        """Forward Home Assistant states updates to ESPHome."""
-        new_state = event.data.get("new_state")
-        if new_state is None:
-            return
-        entity_id = event.data.get("entity_id")
-        await cli.send_home_assistant_state(entity_id, None, new_state.state)
-
     async def _send_home_assistant_state(
-        entity_id: str, new_state: State | None
+        entity_id: str, attribute: str | None, state: State | None
     ) -> None:
         """Forward Home Assistant states to ESPHome."""
-        await cli.send_home_assistant_state(entity_id, None, new_state.state)
+        if state is None or (attribute and attribute not in state.attributes):
+            return
+
+        send_state = state.state
+        if attribute:
+            send_state = state.attributes[attribute]
+            # ESPHome only handles "on"/"off" for boolean values
+            if isinstance(send_state, bool):
+                send_state = "on" if send_state else "off"
+
+        await cli.send_home_assistant_state(entity_id, attribute, str(send_state))
 
     @callback
     def async_on_state_subscription(
         entity_id: str, attribute: str | None = None
     ) -> None:
         """Subscribe and forward states for requested entities."""
+
+        async def send_home_assistant_state_event(event: Event) -> None:
+            """Forward Home Assistant states updates to ESPHome."""
+            await _send_home_assistant_state(
+                event.data["entity_id"], attribute, event.data.get("new_state")
+            )
+
         unsub = async_track_state_change_event(
```

Comment:
Just a note, and it wasn't handled for states before either, but this will send over the attribute even if it hasn't changed.

Reply:
Yeah, I found that odd as well. So, if I create 2 sensors in ESPHome for the same entity, but for different attributes (e.g., a light state & brightness). Change one will trigger an update on both.

As you stated, was already handled for states this way, hence I kept it like that. However, it might be a good time to change the behavior to only communicate changes?

The initial state is synced on the initial connection, so I guess it works if we just send changes? It might be considered a breaking change though.